onnxruntime/core/optimizer/isinf_reducesum_fusion.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/isinf_reducesum_fusion.h"
+
+#include "onnx/defs/attr_proto_util.h"
+
+#include "core/optimizer/initializer.h"
+#include "core/graph/graph_utils.h"
+#include "core/framework/tensorprotoutils.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+/*
+This transform changes the following subgraph pattern:
+Cast --> IsInf --> Cast --> ReduceSum --> Greater
+to
+IsAllFinite --> Not
+*/
+Status IsInfReduceSumFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  std::vector<std::reference_wrapper<Node>> nodes_to_remove;
+  for (auto node_index : node_topology_list) {
+    nodes_to_remove.clear();
+    auto* node_ptr = graph.GetNode(node_index);
+    if (!node_ptr)
+      continue;  // node was removed
+
+    auto& isinf_node = *node_ptr;
+
+    ORT_RETURN_IF_ERROR(Recurse(isinf_node, modified, graph_level, logger));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(isinf_node, "IsInf", {10}) ||
+        isinf_node.GetOutputEdgesCount() != 1 ||
+        !graph.GetNodeOutputsInGraphOutputs(isinf_node).empty()) {
+      continue;
+    }
+
+    std::vector<NodeArg*> input_defs = isinf_node.MutableInputDefs();
+    // see if there is a Cast before IsInf
+    // This will happen if input type is FP16 but IsInf doesnt support fp16, so it will be cast to float/double
+    // This Cast can be skipped as we are replacing the subgraph with IsAllFinite, which supports FP16
+    auto cast1_node_iter = isinf_node.InputNodesBegin();
+    if (cast1_node_iter != isinf_node.InputNodesEnd() &&
+        graph_utils::IsSupportedOptypeVersionAndDomain(*cast1_node_iter, "Cast", {9, 13}) &&
+        cast1_node_iter->GetOutputEdgesCount() == 1) {
+      // check input type of cast node
+      Node& cast1_node = *graph.GetNode(cast1_node_iter->Index());
+      auto cast1_input_defs = cast1_node.MutableInputDefs();
+      auto cast1_input_type = cast1_input_defs[0] ? cast1_input_defs[0]->Type() : nullptr;
+      // remove cast only if the input type is float16
+      if (cast1_input_type != nullptr && (*cast1_input_type) == "tensor(float16)") {
+        input_defs = cast1_input_defs;
+        nodes_to_remove.push_back(cast1_node);
+      }
+    }
+    nodes_to_remove.push_back(isinf_node);
```

Comment:
>nodes_to_remove.push_back(isinf_node); [](start = 4, length = 38)

What if the first "Cast" has other cusomers? Would that fail this optimization?

Reply:
we check for it